Lessen mount lock contention in layer store

kolyshkin opened the pull request:
While testing massive parallel container creation and removal, we see there's a lock contention in layer store that affects container creation performance (in other words, many threads are waiting for the lock). With this patches, container creation time (wall clock time) is about 2x to 3x less than before (measured with both aufs and overlay2 graph drivers).

The (global, per-layerstore) lock in question is held for the duration of
* CreateRWLayer()
* GetRWLayer()
* ReleaseRWLayer()
* CreateRWLayerByGraphID()

**The lock contention observed mostly between two instances of CreateRWLayer()**

Here's a stupid script I am using to test the performance before and after:
```bash
#!/bin/bash

NUMCTS=1000
PARALLEL=100

create() {
	# -v 'vol_{}:/vol'
	time ( seq $NUMCTS | parallel -j$PARALLEL docker create alpine top > /dev/null )
	echo "^^^ CREATE TIME ^^^"
}

remove() {
	time ( docker ps -qa | shuf | tail -n $NUMCTS | parallel -j$PARALLEL docker rm -f '{}' > /dev/null )
	echo "^^^ REMOVE TIME ^^^"
}

# precreate $NUMCTS containers
create
echo

# create another $NUMCTS while removing $NUMCTS containers
create &
remove &
wait

# remove the rest of it
remove
```

Results are:
1. Container creation time (the first timing print by the script) is about 40s before, about 20s after this patchset.

2. Concurrent container creation/removal time (the second and the third timing printed by the script) are about 70s before, 40s after this patchset.

3. Container removal times (last timing printed by the script) stay the same at about 30s.

xinfengliu committed:
layer: optimize layerStore mountL

Goroutine stack analisys shown some lock contention
while doing massively (100 instances of `docker rm`)
parallel image removal, with many goroutines waiting
for the mountL mutex. Optimize it.

With this commit, the above operation is about 3x
faster, with no noticeable change to container
creation times (tested on aufs and overlay2).

kolyshkin@:
- squashed commits
- added description
- protected CreateRWLayer against name collisions by
temporary assiging nil to ls.mounts[name], and treating
nil as "non-existent" in all the other functions.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "layer-mount-lock" git@github.com:kolyshkin/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842358683792
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

kolyshkin commented:
My concern is locking after this patches might not be adequate; perhaps we need to add a per-id (rather than global) lock (using `pkg/locker`) on top of it (but I have yet to look into this).

kolyshkin commented:
> perhaps we need to add a per-id (rather than global) lock (using pkg/locker) on top of it

added (not sure if it's needed though)

kolyshkin commented:
@dmcgowan PTAL

dmcgowan commented:
A finer grain lock is needed in this case. The mounted layer operates on a reference map so there does require locking to prevent alterations on that object in parallel. It maybe possible to just add that lock to `*mountedLayer`, assuming the global lock has already synchronized creation of the same names.

kolyshkin commented:
> A finer grain lock is needed in this case. The mounted layer operates on a reference map so there does require locking to prevent alterations on that object in parallel. It maybe possible to just add that lock to *mountedLayer, assuming the global lock has already synchronized creation of the same names.

@xinfengliu if you have time to address the above comments? I'm currently away for DockerCon and don't have time :(

xinfengliu commented:
@kolyshkin , yes `mountedLayer` need a lock. Sorry I'm on China public holidays (May 1~3) now , I will spend some time on it but I'm not sure I can make it by the end of this week.

kolyshkin committed:
layer: protect mountedLayer.references

Add a mutex to protect concurrent access to mountedLayer.references map.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
@xinfengliu no problem, I think I fixed it.

@dmcgowan PTAL at the last commit (I decided to not use `RWMutex` for the sake of simplicity). Do we need same for `roLayer.references` map?

dmcgowan commented:
> assuming the global lock has already synchronized creation of the same names.

I don't see this part being done. Possibly setting a placeholder in the map and clear it if there is an error. In that case a second call for the same name would get the conflicting error. I agree locking for the entire save is unnecessary just to prevent simultaneous creation of the same name.

kolyshkin commented:
>> assuming the global lock has already synchronized creation of the same names.

> I don't see this part being done. Possibly setting a placeholder in the map and clear it if there is an error. In that case a second call for the same name would get the conflicting error. I agree locking for the entire save is unnecessary just to prevent simultaneous creation of the same name.


@dmcgowan thanks for review! Sorry I missed that part entirely; I think I know what you mean now, please see the updated commit. I had to introduce the third state of `ls.mount[name]` when it's not empty but the value is nil, meaning the name being created. All the other places treat `nil` as non-existent, except for `CreateRWLayer()` which treats it as a name conflict. I'm not in love with this approach as we can have potential nil dereferences but it looks cleaner than yet another lock or map (such as map of names being created).

Oh, if you can think of any test cases that can be created, I'm happy to implement those.

kolyshkin commented on layer/layer_store.go:
```diff
@@ -491,13 +491,20 @@ func (ls *layerStore) CreateRWLayer(name string, parent ChainID, opts *CreateRWL
-	m, ok := ls.mounts[name]
-	if ok {
+	if _, ok := ls.mounts[name]; ok {
 		return nil, ErrMountNameConflict
```
I forgot to call Unlock on this path :-\ fix is coming

kolyshkin replied:
fixed now
>+		ls.mountL.Unlock()
￼

kolyshkin commented:
Some changes based on @dmcgowan and my own review:
 - added locking to loadMount() (just for consistency/uniformity, techinically it's not required in there);
 - added missing Unlock() to an error path in CreateRWLayer;
 - simplified return statements in CreateRWLayer as we use named return values;
 - simplified map element presence check by dropping `ok` assignment and check (`ok && m != nil` --> `m != nil`).

dmcgowan reviewed (approved):
LGTM

kolyshkin commented:
A comment from @tonistiigi :

> have some concerns about https://github.com/moby/moby/pull/39135. eg.in https://github.com/moby/moby/pull/39135/files#diff-5137c87d28c11ce0691d2f12e124409eR587 deleteRef, hasRef + remove is not atomic anymore. so if you call GetRWLayer after hasReference you get a rwlayer but the storage gets deleted in another goroutine

is to be addressed in a followup
